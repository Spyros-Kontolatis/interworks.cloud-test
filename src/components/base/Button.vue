<template>
  <button class="border border-primary relative" :class="[...buttonClasses]">
    <Loader v-if="loading" class="w-8 h-8 top-3 left-2" />

    <slot></slot>
  </button>
</template>

<script setup lang="ts">
import { defineProps, computed } from "vue";
import {
  sizes,
  textColors,
  bgColors,
  rounded,
  outline,
} from "@/config/components/base/button";
import type { ButtonProps } from "@/types/components/base/button";
import Loader from "./Loader.vue";

const props = defineProps<ButtonProps>();

const buttonClasses = computed(() => {
  return [
    bgColors[props.color],
    textColors[props.color],
    sizes[props.size],
    props.rounded && rounded[props.size],
    props.outline && outline[props.color],
    (props.disabled || props.loading) &&
      "opacity-50 cursor-not-allowed pointer-events-none",
  ];
});
</script>
