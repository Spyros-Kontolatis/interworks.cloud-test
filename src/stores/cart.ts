import { defineStore } from "pinia";
import { ref } from "vue";
import type { Cart, CartItem } from "@/types/stores/cart/cart";
import Service from "@/services/Service";

export const useCartStore = defineStore("cart", () => {
  const cart = ref<Cart>({
    id: "",
    items: [],
    totalPrice: 0,
    totalQty: 0,
  });
  const addToCart = (item: CartItem) => {
    const isItemPresent = cart.value.items.find(
      ({ product: { id } }) => id === item.product.id
    );
    if (isItemPresent) {
      isItemPresent.quantity++;
    } else {
      cart.value.items.push(item);
    }
    cart.value.totalQty++;
    cart.value.totalPrice += item.product.price;
    if (cart.value.id) {
      Service.cart().update(cart.value.id, cart.value).execute();
    } else {
      cart.value.id = Date.now().toString();
      Service.cart().create(cart.value).execute();
    }
  };

  const updateItemQuantity = (itemId: string, mode: "remove" | "add") => {
    const itemInCart = cart.value.items.find(
      ({ product: { id } }) => id === itemId
    );
    if (itemInCart) {
      switch (mode) {
        case "add":
          itemInCart.quantity++;
          cart.value.totalQty++;
          cart.value.totalPrice += itemInCart.product.price;
          break;
        case "remove":
          itemInCart.quantity--;
          cart.value.totalQty--;
          cart.value.totalPrice -= itemInCart.product.price;
          break;
      }
    }
    Service.cart().update(cart.value.id, cart.value).execute();
  };

  const removeItemFromCart = (item: CartItem) => {
    cart.value.items = cart.value.items.filter(
      ({ product: { id } }) => id !== item.product.id
    );

    cart.value.totalQty -= item.quantity;
    cart.value.totalPrice -= item.product.price * item.quantity;
    Service.cart().update(cart.value.id, cart.value).execute();
  };
  return {
    cart,
    removeItemFromCart,
    addToCart,
    updateItemQuantity,
  };
});
