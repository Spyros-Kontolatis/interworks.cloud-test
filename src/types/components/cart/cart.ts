import type { CartItem } from "@/types/stores/cart/cart";

export interface CartItemProps {
  item: CartItem;
}
