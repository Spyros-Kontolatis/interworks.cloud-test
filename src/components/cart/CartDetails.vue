<template>
  <div class="border border-gray-400 p-4 rounded-md ml-4">
    <div class="flex flex-col">
      <h4 class="font-bold text-start" v-if="cart.items.length">Products:</h4>
      <div
        class="flex justify-between"
        v-for="{ product: { name, id, price }, quantity } in cart.items"
        :key="id"
      >
        <div class="flex grow">
          {{ name }}
          <div class="text-xs flex items-center ml-2">
            ({{ quantity }} x <PriceItem :price="price" />)
          </div>
        </div>
        <div class="text-xs flex gap-2">
          <PriceItem :price="price * quantity" />
        </div>
      </div>
      <div
        class="flex justify-between font-extrabold border-t border-t-gray-400 pt-2"
      >
        <div class="">TOTAL:</div>
        <div>
          <PriceItem :price="cart.totalPrice" />
        </div>
      </div>
    </div>
  </div>
</template>

<script setup lang="ts">
import { useCartStore } from "@/stores/cart";
import { storeToRefs } from "pinia";
import PriceItem from "../product/PriceItem.vue";

const cartStore = useCartStore();
const { cart } = storeToRefs(cartStore);
</script>
