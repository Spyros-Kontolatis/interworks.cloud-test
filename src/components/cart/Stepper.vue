<template>
  <div class="flex">
    <Button
      size="sm"
      color="primary"
      outline
      rounded
      :disabled="step === 1 || loading"
      class="rounded-r-none"
      @click="$emit('update:step', step - 1, 'remove')"
      ><div class="font-bold text-lg">-</div></Button
    >
    <div class="border border-primary py-2 px-4">{{ step }}</div>
    <Button
      size="sm"
      color="primary"
      rounded
      :disabled="loading"
      @click="$emit('update:step', step + 1, 'add')"
      class="rounded-l-none"
      ><div class="font-bold text-lg">+</div></Button
    >
  </div>
</template>

<script setup lang="ts">
import { StepperProps } from "@/types/components/cart/stepper";
import Button from "../base/Button.vue";
import { defineProps, defineEmits } from "vue";

defineProps<StepperProps>();
defineEmits(["update:step"]);
</script>
