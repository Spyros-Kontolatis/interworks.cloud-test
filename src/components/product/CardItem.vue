<template>
  <div class="bg-light-gray rounded-2xl flex flex-col">
    <img class="h-2/3 rounded-t-2xl cursor-pointer" :src="product.image" />
    <div class="px-4">
      <p class="mt-2">{{ product.name }}</p>
      <p class="my-2">{{ product.description }}</p>
      <PriceItem :price="product.price" class="my-2 grow" />
      <Button
        color="primary"
        size="xl"
        rounded
        :loading="loading"
        class="mb-2 w-full"
        @click="addItemToCart"
        >Add To Cart</Button
      >
    </div>
  </div>
</template>

<script setup lang="ts">
import { defineProps, PropType, ref } from "vue";
import { useCartStore } from "@/stores/cart";
import PriceItem from "./PriceItem.vue";
import Button from "../base/Button.vue";
import type { Product } from "@/types/components/products/product";

const { addToCart } = useCartStore();
const loading = ref(false);
const props = defineProps({
  product: { type: Object as PropType<Product>, required: true },
});

const addItemToCart = () => {
  loading.value = true;
  addToCart({ product: props.product, quantity: 1 });
  loading.value = false;
};
</script>
