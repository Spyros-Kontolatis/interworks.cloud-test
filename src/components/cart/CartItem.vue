<template>
  <div class="border border-gray-400 relative p-4 rounded-md mr-4">
    <div
      v-if="loading"
      class="w-full h-40 absolute top-0 left-0 bg-gray-200 opacity-40"
    >
      <Loader class="w-10 h-10 top-1/3" />
    </div>
    <div class="flex gap-4">
      <div>
        <img :src="item.product.image" width="200" />
      </div>
      <div class="flex grow justify-between text-start">
        <div class="flex flex-col gap-2">
          <div class="font-extrabold">{{ item.product.name }}</div>
          <div>{{ item.product.description }}</div>
          <Stepper
            :loading="loading"
            v-model:step="step"
            @update:step="(_newVal, mode) => updateQuantity(mode)"
          />
        </div>
        <div class="mr-8 flex flex-col">
          <PriceItem :price="item.product.price * item.quantity" />
          <img
            width="20"
            src="../../assets/icons/trash.svg"
            class="cursor-pointer"
            @click="removeItemFromCart(item)"
          />
        </div>
      </div>
    </div>
  </div>
</template>

<script setup lang="ts">
import { defineProps, ref } from "vue";
import type { CartItemProps } from "@/types/components/cart/cart";
import { useCartStore } from "@/stores/cart";
import PriceItem from "../product/PriceItem.vue";
import Loader from "../base/Loader.vue";
import Stepper from "./Stepper.vue";

const props = defineProps<CartItemProps>();
const { updateItemQuantity, removeItemFromCart } = useCartStore();
const step = ref(props.item.quantity);
const loading = ref(false);

const updateQuantity = (mode: "remove" | "add") => {
  loading.value = true;
  updateItemQuantity(props.item.product.id, mode);
  loading.value = false;
};
</script>
