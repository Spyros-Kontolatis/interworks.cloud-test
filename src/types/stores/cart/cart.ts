import { Product } from "../../components/products/product";

export type CartItem = {
  product: Product;
  quantity: number;
};

export type Cart = {
  items: Array<CartItem>;
  totalQty: number;
  totalPrice: number;
};
