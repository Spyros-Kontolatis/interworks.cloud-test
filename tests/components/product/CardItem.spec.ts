import { shallowMount } from "@vue/test-utils";
import CardItem from "../../../src/components/product/CardItem.vue";
import { setActivePinia, createPinia } from "pinia";
import { createTestingPinia } from "@pinia/testing";
import { useCartStore } from "@/stores/cart";
import Button from "@/components/base/Button.vue";
beforeEach(() => {
  setActivePinia(createPinia());
});

describe("CardItem.vue", () => {
  it("renders the component", () => {
    const wrapper = shallowMount(CardItem, {
      props: {
        product: {
          name: "Product 1",
          price: 1,
          description: "A product",
          image: "",
          id: "1",
        },
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });
  it("adds product to cart", async () => {
    const wrapper = shallowMount(CardItem, {
      props: {
        product: {
          name: "Product 1",
          price: 1,
          description: "A product",
          image: "",
          id: "1",
        },
      },
      global: {
        plugins: [
          createTestingPinia({
            initialState: {
              cart: {
                cart: {
                  items: [
                    {
                      product: {
                        name: "Product 2",
                        price: 1,
                        description: "A product",
                        image: "",
                        id: "2",
                      },
                      quantity: 1,
                    },
                  ],
                  totalQty: 1,
                  totalPrice: 1,
                },
              },
            },
            stubActions: false,
          }),
        ],
      },
    });
    const cartStore = useCartStore();
    expect(wrapper.element).toMatchSnapshot();
    const addToCartButton = wrapper.findComponent(Button);
    addToCartButton.vm.$emit("click");
    await wrapper.vm.$nextTick();
    expect(cartStore.cart.items).toHaveLength(2);
  });
});
