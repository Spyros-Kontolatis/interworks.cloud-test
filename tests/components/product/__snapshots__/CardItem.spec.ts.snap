// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardItem.vue adds product to cart 1`] = `
<div
  class="bg-light-gray rounded-2xl flex flex-col"
>
  <!--v-if-->
  <div
    class="px-4"
  >
    <p
      class="mt-2"
    >
      Product 1
    </p>
    <p
      class="my-2"
    >
      A product
    </p>
    <price-item-stub
      class="my-2 grow"
      price="1"
    />
    <button-stub
      class="mb-2 w-full"
      color="primary"
      disabled="false"
      loading="false"
      outline="false"
      rounded="true"
      size="xl"
    />
  </div>
</div>
`;

exports[`CardItem.vue renders the component 1`] = `
<div
  class="bg-light-gray rounded-2xl flex flex-col"
>
  <!--v-if-->
  <div
    class="px-4"
  >
    <p
      class="mt-2"
    >
      Product 1
    </p>
    <p
      class="my-2"
    >
      A product
    </p>
    <price-item-stub
      class="my-2 grow"
      price="1"
    />
    <button-stub
      class="mb-2 w-full"
      color="primary"
      disabled="false"
      loading="false"
      outline="false"
      rounded="true"
      size="xl"
    />
  </div>
</div>
`;
